### DISCUSSION TITLE
Docker image version update changed digest arch from amd64 to i386

### DISCUSSION THREAD

**Discussion by LeoniePhiline** [renovatebot/renovate#22639] [Request Help] [answered]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

v35.110.0

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the problem

# History

Renovate-bot, v35.110.0 at the time, changed the image architecture in a digest update from `amd64` to `i386`.

The previous image version was `rust:1.69.0-slim-bullseye@sha256:caba086dac32589a2c4e5ac43212eb2fa2eb1ddb49b19dbd976c89115af56f70`, which is amd64:

```bash
docker run --rm rust:1.69.0-slim-bullseye@sha256:caba086dac32589a2c4e5ac43212eb2fa2eb1ddb49b19dbd976c89115af56f70 dpkg --print-architecture
amd64
```

- On Jun 3, 2023 6:56am GMT+0200, renovate-bot mentioned the update in the dependency dashboard.
- On Jun 3, 2023 8:57am GMT+0200, the following diff was merged as "chore(deps): update rust docker tag to v1.70.0"

This MR contains the following updates:

| Package | Update | Change |
|---|---|---|
| rust | minor | `1.69.0-slim-bullseye` -> `1.70.0-slim-bullseye` |

```diff
@@ -1,16 +1,16 @@
 ARG DEPENDENCY_PROXY_GROUP_IMAGE_PREFIX
 
-FROM $DEPENDENCY_PROXY_GROUP_IMAGE_PREFIX/rust:1.69.0-slim-bullseye@sha256:caba086dac32589a2c4e5ac43212eb2fa2eb1ddb49b19dbd976c89115af56f70
+FROM $DEPENDENCY_PROXY_GROUP_IMAGE_PREFIX/rust:1.70.0-slim-bullseye@sha256:241f150757990575e5e157c2d4af33d914a6c8dc6b9a1593ca3b86dab27c9ae6
```

Where  `rust:1.70.0-slim-bullseye@sha256:241f150757990575e5e157c2d4af33d914a6c8dc6b9a1593ca3b86dab27c9ae6` was in fact `i386`:

```bash
docker run --rm -ti rust:1.70.0-slim-bullseye@sha256:241f150757990575e5e157c2d4af33d914a6c8dc6b9a1593ca3b86dab27c9ae6 dpkg --print-architecture
i386
```

## Relevant log

```
DEBUG: packageFiles with updates (repository=applications/releasectl, baseBranch=main)
"config": {
  "regex": [
    {
      "datasourceTemplate": "docker",
      "deps": [
        {
          "currentDigest": "sha256:caba086dac32589a2c4e5ac43212eb2fa2eb1ddb49b19dbd976c89115af56f70",
          "currentValue": "1.69.0-slim-bullseye",
          "currentVersion": "1.69.0",
          "datasource": "docker",
          "depName": "rust",
          "fixedVersion": "1.69.0-slim-bullseye",
          "isSingleVersion": true,
          "packageName": "rust",
          "registryUrl": "https://index.docker.io",
          "replaceString": "$DEPENDENCY_PROXY_GROUP_IMAGE_PREFIX/rust:1.69.0-slim-bullseye@sha256:caba086dac32589a2c4e5ac43212eb2fa2eb1ddb49b19dbd976c89115af56f70",
          "versioning": "docker",
          "warnings": [],
          "updates": [
            {
              "bucket": "non-major",
              "newVersion": "1.70.0",
              "newValue": "1.70.0-slim-bullseye",
              "newMajor": 1,
              "newMinor": 70,
              "updateType": "minor",
              "newDigest": "sha256:241f150757990575e5e157c2d4af33d914a6c8dc6b9a1593ca3b86dab27c9ae6",
              "branchName": "renovate/rust-1.x"
            },
            {
              "updateType": "digest",
              "newValue": "1.69.0-slim-bullseye",
              "newDigest": "sha256:2522978d04d670d70389ee2d91ae7d266f622eb619560f6a242ee4c5544a39c5",
              "branchName": "renovate/rust-1.69.0-slim-bullseye"
            }
          ]
        }
      ],
      "matchStrings": [
        "\\$\\{?(?:CI_)?DEPENDENCY_PROXY_(?:DIRECT_)?GROUP_IMAGE_PREFIX\\}?\\/(?<depName>[^:]+):(?<currentValue>[a-z0-9.-]+)(?:@(?<currentDigest>sha256:[a-f0-9]+))?"
      ],
      "packageFile": "ci/base.Dockerfile",
      "versioningTemplate": "docker"
    }
  ]
}
```

The problematic line: `"newDigest": "sha256:241f150757990575e5e157c2d4af33d914a6c8dc6b9a1593ca3b86dab27c9ae6",`

# Impact

This means that renovate-bot changed the architecture, which caused issues further down the line: The image is the base for rust binary compilation (`cargo build --release`), which resulted in a 32 bit binary linking to a non-existent ELF interpreter:

```
readelf -a bin/releasectl
ELF Header:
  Magic:   7f 45 4c 46 01 01 01 00 00 00 00 00 00 00 00 00 
  Class:                             ELF32
[...]

file bin/releasectl
bin/releasectl ELF 32-bit LSB pie executable, Intel 80386, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux.so.2, BuildID[sha1]=db1fde118ab2b30dab3a528c05ac92b9dcc867b6, for GNU/Linux 3.2.0, with debug_info, not stripped
```

The ELF interpreter `/lib/ld-linux.so.2` is linked, which does not even exist on the build system, neither on the 64 bit target systems where the binary is ot be run.

Expected `file` output would have been:

```
file bin/releasectl
bin/releasectl ELF 64-bit LSB pie executable, x86-64, version 1 (SYSV), dynamically linked, interpreter /lib64/ld-linux-x86-64.so.2, BuildID[sha1]=d3b2ed501812d5fc30e7f92cc71f2438e66f6a31, for GNU/Linux 3.2.0, with debug_info, not stripped
```

Where `/lib64/ld-linux-x86-64.so.2` is the correct 64bit ELF interpreter.

# Follow-up

Interestingly on the next digest update, renovate-bot switched back to `amd64` with "chore(deps): update rust:1.70.0-slim-bullseye docker digest to 585eedd":

This MR contains the following updates:

| Package | Update | Change |
|---|---|---|
| rust | digest | `241f150` -> `585eedd` |

```diff
@@ -1,16 +1,16 @@
 ARG DEPENDENCY_PROXY_GROUP_IMAGE_PREFIX
 
-FROM $DEPENDENCY_PROXY_GROUP_IMAGE_PREFIX/rust:1.70.0-slim-bullseye@sha256:241f150757990575e5e157c2d4af33d914a6c8dc6b9a1593ca3b86dab27c9ae6
+FROM $DEPENDENCY_PROXY_GROUP_IMAGE_PREFIX/rust:1.70.0-slim-bullseye@sha256:585eeddab1ec712dade54381e115f676bba239b1c79198832ddda397c1f58bfc
```

where `rust:1.70.0-slim-bullseye@sha256:585eeddab1ec712dade54381e115f676bba239b1c79198832ddda397c1f58bfc` is `amd64`:

```
docker run --rm -ti rust:1.70.0-slim-bullseye@sha256:585eeddab1ec712dade54381e115f676bba239b1c79198832ddda397c1f58bfc dpkg --print-architecture
amd64
```

# Speculation

I could imagine the `amd64` image was not available on docker hub at the time renovate-bot created the MR for updating from `rust:1.69.0-slim-bullseye` to `rust:1.70.0-slim-bullseye`, thus choosing the `i386` as fallback.

This can cause severe issues down the line, with full production systems failing, as the newly deployed binaries cannot find the linked 32bit ELF interpreter.

I find it paramount to keep the architecture when updating digests and never fall back to other architectures.

### Relevant debug logs

See above.

### Have you created a minimal reproduction repository?

Impossible to reproduce, since I cannot restore the available digests on docker hub at the time.

**Comment by rarkins** [comment]
Consider using amd64/ prefixes for library images. e.g. amd64/rust

**Reply by LeoniePhiline** [reply]
Thanks! 

Is this considered a solution or rather a workaround?

(Considering OCI indexes are, as I understand, the new solution for this old problem. See also https://github.com/docker-library/official-images#architectures-other-than-amd64)

**Reply by rarkins** [reply]
If you have a combined architecture image/tag then renovate should pick the shared digest and not anything architecture-specific. But it's very hard to know what happened here. We use amd64 prefixes for library images ourselves to remove this Docker publish race condition  but at the same time any behavior like you describe is not intended

**Reply by LeoniePhiline** [reply]
Thank you! I'll mark this as answer.

Might it nevertheless be possible (meaning: a desirable new feature) to **check the previous digest's architecture and restrict the new digest to the same architecture**?

As far as I can see in the UI, there isn't a "shared digest", but only a digest per architecture. Example: https://hub.docker.com/layers/library/rust/1.70.0-slim-bullseye/images/sha256-53f49ce140a3d19c775da59c3d9b124c1830425ef5568ca45d1f29e76cb15f96?context=explore 

The `amd64` arch is traditionally the default, AFAIK.

**Reply by rarkins** [reply]
@viceice ?

**Reply by Churro** [reply]
renovate doesn't switch to another architecture, it falls back to the manifest digest if no update could be retrieved for the provided arch. It may happen though that e.g., the manifest with the new digest includes no amd64 image yet, which could explain why docker picked i386.

`caba086dac32589a2c4e5ac43212eb2fa2eb1ddb49b19dbd976c89115af56f70` is the **manifest digest** of `rust:1.69.0-slim-bullseye` ([see here](https://hub.docker.com/layers/library/rust/1.69.0-slim-bullseye/images/sha256-caba086dac32589a2c4e5ac43212eb2fa2eb1ddb49b19dbd976c89115af56f70)), not the amd64-specific digest. You should be able to cross-check this by inspecting the log. If the digest is architecture-specific the debug log will include the line `Current digest ... relates to architecture amd64`.

**Reply by viceice** [reply, marked as answer]
@Churro is right. that's the current renovate behavior. maybe we should somehow cancel the update when we've an arch specific digest and the update doesn't have one yet.

**Reply by LeoniePhiline** [reply]
That sounds like a good solution. 

> maybe we should somehow cancel the update when we've an arch specific digest and the update doesn't have one yet.

@viceice Should I keep this discussion marked as resolved, or rather unmark to ensure it is not forgotten?

**Reply by viceice** [reply]
can someone open a bug issue for docker datasource?

it should not fallback to manifest list digest if old digest points to an arch image and that isn't found in new manifest list.

**Reply by rarkins** [reply]
How/where could Renovate achieve that though? it's just a sha string in the repo without annotation of architecture, so Renovate would need to do additional querying for every single dependency, wouldn't it?

I'm concerned that this is actually a complex feature with API rate limit implications, and not a some simple bug fix.

**Reply by viceice** [reply]
renovate already does fetch the arch of the old Sha and tries to find a matching arch Sha for the new version. if it finds it in the new manifest everything is fine, but if the arch is not in the new manifest, renovate falls back to the manifest sha, which shouldn't be done.

**Reply by LeoniePhiline** [reply]
Do you know if the solution ("maybe we should somehow cancel the update when we've an arch specific digest and the update doesn't have one yet") has been implemented in the mean time?

I was unable to find anything, but I might have overlooked something.
